### PR TITLE
Api 1.7

### PR DESCRIPTION
**Pull request by vizakhar**
API promotion

### CODE REVIEW

File: bin/pbiviz-new.js [hunk 1]
```diff
@@ -62,10 +62,19 @@ if (program.force) {
     ConsoleWriter.warn('Running with force flag. Existing files will be overwritten');
 }
 
+let templateExternalJS = {
+    defalut: [],
+    rhtml: [],
+    rvisual: [],
+    slicer: ["node_modules/powerbi-models/dist/models.js", "node_modules/powerbi-visuals-utils-dataviewutils/lib/index.js"],
```
Comment (AviSander) [inline bin/pbiviz-new.js]: should not reference node_modules. see other visuals for cleaner way.

File: templates/visuals/slicer/package.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "visual",
-  "devDependencies": {
+  "dependencies": {
+	"powerbi-models": "0.10.0",
     "powerbi-visuals-utils-dataviewutils": "1.0.1"
```
Comment (AviSander) [inline templates/visuals/slicer/package.json]: please add tslints

File: bin/pbiviz-new.js [hunk 3]
```diff
@@ -62,10 +62,19 @@ if (program.force) {
     ConsoleWriter.warn('Running with force flag. Existing files will be overwritten');
 }
 
+let templateExternalJS = {
```
Comment (AviSander) [inline bin/pbiviz-new.js]: we need to avoid such 'hacks'. let's get this working the right way with modular json files managing it per template.

File: lib/VisualGenerator.js [hunk 4]
```diff
@@ -240,7 +248,7 @@ class VisualGenerator {
 
                     copyVisualTemplate(visualPath, buildOptions.template);
                     generateSettings(visualPath);
-                    createPbiVizJson(visualPath, visualOptions);
+                    createPbiVizJson(visualPath, visualOptions, buildOptions.externalJS);
```
Comment (AviSander) [inline lib/VisualGenerator.js]: like before - let's not add buildOptions.externalJS but get it working with pbiviz.json per template

File: templates/visuals/slicer/package.json [hunk 5]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "visual",
-  "devDependencies": {
+  "dependencies": {
+	"powerbi-models": "0.10.0",
```
Comment (AviSander) [inline templates/visuals/slicer/package.json]: tabs\spaces

File: templates/visuals/slicer/pbiviz.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{
+  "visual": {
+    "name": "slicer",
+    "displayName": "slicer",
```
Comment (AviSander) [inline templates/visuals/slicer/pbiviz.json]: slicerSample

File: templates/visuals/slicer/pbiviz.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    "guid": "slicerADBF8B769AEF46CBAEA9AA1779FA9885",
```
Comment (AviSander) [inline templates/visuals/slicer/pbiviz.json]: slicerSampleADBF8B769AEF46CBAEA9AA1779FA9885

File: templates/visuals/slicer/pbiviz.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    "visualClassName": "Visual",
```
Comment (AviSander) [inline templates/visuals/slicer/pbiviz.json]: slicerVisual

File: templates/visuals/slicer/pbiviz.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (AviSander) [inline templates/visuals/slicer/pbiviz.json]: why is this file required? please remove.

File: templates/visuals/slicer/tsconfig.json [hunk 10]
```diff
@@ -8,8 +8,9 @@
         "out": "./.tmp/build/visual.js"
     },
     "files": [
-        ".api/v1.5.0/PowerBI-visuals.d.ts",
+        ".api/v1.7.0/PowerBI-visuals.d.ts",
         "node_modules/powerbi-visuals-utils-dataviewutils/lib/index.d.ts",
+		"node_modules/powerbi-models/dist/models-noexports.d.ts",
```
Comment (AviSander) [inline templates/visuals/slicer/tsconfig.json]: spaces\tabs

File: templates/visuals/.api/v1.7.0/schema.stringResources.json [hunk 11]
```diff
@@ -0,0 +1,67 @@
+{
+    "PBI_API_VERSION": "v1.6.0",
```
Comment (AviSander) [inline templates/visuals/.api/v1.7.0/schema.stringResources.json]: v1.7.0

File: templates/visuals/.api/v1.7.0/schema.pbiviz.json [hunk 12]
```diff
@@ -0,0 +1,103 @@
+{
+    "PBI_API_VERSION": "v1.6.0",
```
Comment (AviSander) [inline templates/visuals/.api/v1.7.0/schema.pbiviz.json]: v1.7.0

File: templates/visuals/.api/v1.7.0/schema.dependencies.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
+{
+    "PBI_API_VERSION": "v1.6.0",
```
Comment (AviSander) [inline templates/visuals/.api/v1.7.0/schema.dependencies.json]: v1.7.0

File: certs/PowerBICustomVisualTest_private.key [hunk 14]
```diff
@@ -1,28 +1,34 @@
+Bag Attributes
```
Comment (AviSander) [inline certs/PowerBICustomVisualTest_private.key]: please remove these files (Rebase)

File: certs/PowerBICustomVisualTest_public.crt [hunk 15]
```diff
@@ -1,22 +1,24 @@
+Bag Attributes
```
Comment (AviSander) [inline certs/PowerBICustomVisualTest_public.crt]: please remove this file (Rebase)

File: lib/VisualGenerator.js [hunk 16]
```diff
@@ -72,11 +72,47 @@ function generateVisualName(displayName) {
...
+    // read the global template data
+    // and generate the actual file content
+    let globalTemplateData = fs.readFileSync(globalTemplatePath);
+    let data = _.template(globalTemplateData)(options);
+
+    // write out the target file content
+    let targetPath = path.join(visualPath, 'pbiviz.json');
+    fs.writeFileSync(targetPath, data);   
+
+    let templatePath = path.join(VISUAL_TEMPLATES_PATH, templateName);
```
Comment (AviSander) [inline lib/VisualGenerator.js]: nice.

File: lib/VisualGenerator.js [hunk 17]
```diff
@@ -204,7 +240,8 @@ function validTemplate(templateName) {
 const defaultOptions = {
     force: false,
     template: 'default',
-    apiVersion: API_VERSION
+    apiVersion: API_VERSION,
+    externalJS: []
```
Comment (AviSander) [inline lib/VisualGenerator.js]: still required?

File: templates/visuals/slicer/pbiviz.json [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  "visual": {
+    "name": "slicerSample",
```
Comment (AviSander) [inline templates/visuals/slicer/pbiviz.json]: are these names required? 'sample'? guid? etc?

File: templates/visuals/slicer/tslint.json [hunk 19]
```diff
@@ -0,0 +1,79 @@
+{
```
Comment (AviSander) [inline templates/visuals/slicer/tslint.json]: this file should be shared among all templates - in 'common'

File: templates/visuals/slicer/package.json (hunk 2 above)
Comment (vizakhar) [inline templates/visuals/slicer/package.json]: done

File: bin/pbiviz-new.js (hunk 3 above)
Comment (vizakhar) [inline bin/pbiviz-new.js]: done

File: certs/PowerBICustomVisualTest_private.key (hunk 14 above)
Comment (vizakhar) [inline certs/PowerBICustomVisualTest_private.key]: I merged master into the branch => these files are updated

File: templates/visuals/slicer/pbiviz.json (hunk 18 above)
Comment (vizakhar) [inline templates/visuals/slicer/pbiviz.json]: done removing everything other than externalJS{}

File: lib/VisualGenerator.js (hunk 17 above)
Comment (vizakhar) [inline lib/VisualGenerator.js]: yes. still required.

File: lib/VisualGenerator.js (hunk 16 above)
Comment (vizakhar) [inline lib/VisualGenerator.js]: thx

File: certs/PowerBICustomVisualTest_public.crt (hunk 15 above)
Comment (vizakhar) [inline certs/PowerBICustomVisualTest_public.crt]: files updated.

File: certs/PowerBICustomVisualTest_private.key (hunk 14 above)
Comment (vizakhar) [inline certs/PowerBICustomVisualTest_private.key]: merged master into the branch => the certificates are up-to-date

**Comment by vizakhar** [comment]
applied the requested changes. Npm tests have all passed.
